### PR TITLE
Add support for ALPN

### LINKED ISSUE #21: Support NPN / ALPN

### PR DESCRIPTION
**Pull request by bifurcation**
Fixes #21 

r? @ekr

### CODE REVIEW

File: conn.go [hunk 1]
```diff
@@ -66,10 +66,12 @@ type Config struct {
 	CipherSuites        []cipherSuite
 	Groups              []namedGroup
 	SignatureAlgorithms []signatureAndHashAlgorithm
+	NextProtos          []string
```
Comment (ekr) [inline conn.go]: I feel like this should be an array of bytes instead.

File: conn.go [hunk 2]
```diff
@@ -523,6 +529,11 @@ func (c *Conn) clientHandshake() error {
 	sa := signatureAlgorithmsExtension{algorithms: c.config.SignatureAlgorithms}
 	dv := draftVersionExtension{version: draftVersionImplemented}
 
+	var alpn *alpnExtension
+	if (c.config.NextProtos != nil) && (len(c.config.NextProtos) > 0) {
```
Comment (ekr) [inline conn.go]: Can this kind of vector be nil?

File: conn.go [hunk 3]
```diff
@@ -851,6 +873,19 @@ func (c *Conn) serverHandshake() error {
 		}
 	}
 
+	// Find early_data extension and handle early data
```
Comment (ekr) [inline conn.go]: This comment seems wrong

File: conn.go [hunk 4]
```diff
@@ -851,6 +873,19 @@ func (c *Conn) serverHandshake() error {
...
+	sendALPN := false
+	var serverALPN *alpnExtension
+	if gotALPN {
+		for _, proto := range clientALPN.protocols {
+			if c.config.enabledProto[proto] {
```
Comment (ekr) [inline conn.go]: I would tend to prefer searching through the list here rather than building an unnecessary map.

File: conn.go [hunk 5]
```diff
@@ -1061,6 +1096,13 @@ func (c *Conn) serverHandshake() error {
 			return err
 		}
 	}
+	if sendALPN {
```
Comment (ekr) [inline conn.go]: Why can't you just add the extension above or at least use serverALPN as the conditional

File: extensions.go [hunk 6]
```diff
@@ -567,6 +567,56 @@ func (ed *earlyDataExtension) Unmarshal(data []byte) (int, error) {
...
+
+func (alpn alpnExtension) Type() helloExtensionType {
+	return extensionTypeALPN
+}
+
+func (alpn alpnExtension) Marshal() ([]byte, error) {
+	listData := []byte{}
+	for _, proto := range alpn.protocols {
+		listData = append(listData, byte(len(proto)))
+		listData = append(listData, []byte(proto)...)
```
Comment (ekr) [inline extensions.go]: I don't know go well, what does the ... do?

File: extensions.go [hunk 7]
```diff
@@ -567,6 +567,56 @@ func (ed *earlyDataExtension) Unmarshal(data []byte) (int, error) {
...
+	}
+
+	listLen := len(listData)
+	lenData := []byte{byte(listLen >> 8), byte(listLen)}
+	return append(lenData, listData...), nil
```
Comment (ekr) [inline extensions.go]: You can't append integers directly?

File: extensions.go [hunk 8]
```diff
@@ -567,6 +567,56 @@ func (ed *earlyDataExtension) Unmarshal(data []byte) (int, error) {
...
+	if len(data) < 2+listLen {
+		return 0, fmt.Errorf("tls.alpn: Too short for proto list")
+	}
+
+	read := 2
+	alpn.protocols = []string{}
+	for read < listLen+2 {
+		itemLen := int(data[read])
+		if 2+listLen < read+itemLen {
+			return 0, fmt.Errorf("tls.alpn: Too short for list")
```
Comment (ekr) [inline extensions.go]: Isn't this too long?

File: extensions.go [hunk 9]
```diff
@@ -567,6 +567,56 @@ func (ed *earlyDataExtension) Unmarshal(data []byte) (int, error) {
...
+	listLen := (int(data[0]) << 8) + int(data[1])
+	if len(data) < 2+listLen {
+		return 0, fmt.Errorf("tls.alpn: Too short for proto list")
+	}
+
+	read := 2
+	alpn.protocols = []string{}
+	for read < listLen+2 {
+		itemLen := int(data[read])
+		if 2+listLen < read+itemLen {
```
Comment (ekr) [inline extensions.go]: Why not increment read here?

**Comment by ekr** [comment]
This looks basically right module the string/byte array issues noted above. I notice you have no functional tests for alpn match/mismatch

File: conn.go (hunk 1 above)
Comment (bifurcation) [inline conn.go]: `crypto/tls` has it as `[]string`, so I'm going to keep it to be consistent.

File: conn.go (hunk 2 above)
Comment (bifurcation) [inline conn.go]: Yep.  Technically, it's a "slice" vs. an "array", so it can be nil.

File: extensions.go (hunk 6 above)
Comment (bifurcation) [inline extensions.go]: That's how you append another slice, vs. appending a single element (as on the previous line).

**Pull request merged**